### DISCUSSION TITLE
Typescript validation

### DISCUSSION THREAD

**Discussion by ebisbe** [ChristopheBougere/asl-validator#146] [Ideas]
We are trying to add Typescript types to https://github.com/serverless-operations/serverless-step-functions/pull/585 and that package is using asl-validator to validate before deploy. 

Could be something interesting having all the types definitions ( mostly for the definition state language ) ? 

@zirkelc

**Comment by ChristopheBougere** [comment]
These are the types used by asl-validator:
https://github.com/ChristopheBougere/asl-validator/blob/main/src/types.ts

I'm not sure if they are available from outside, would it help if they were?

**Reply by ebisbe** [reply]
I saw that file but that's not what I'm looking for. Types for Map, Pass, Choice, etc. How do you validate https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-stepfunctions-statemachine.html#cfn-stepfunctions-statemachine-definition is correct?

**Reply by ChristopheBougere** [reply]
oh, state machines and states are validated with JSON schemas.
See here the schema for the state machine: https://github.com/ChristopheBougere/asl-validator/blob/main/src/schemas/state-machine.json
And for the choice state for instance: https://github.com/ChristopheBougere/asl-validator/blob/main/src/schemas/choice.json
We don't have more precise TypeScript types that the ones in  https://github.com/ChristopheBougere/asl-validator/blob/main/src/types.ts unfortunately

**Reply by ebisbe** [reply]
ok, maybe we can use the json files. I think another project was converting jscon schemas to types. I'll have to dig a litter bit more.

**Reply by massfords** [reply]
The json files do not encode all the rules for the language. This is due to limits in the JSON schema as well as some choices made to improve error reporting (see #119). The `Map` state JSON schema for example doesn't encode that ItemSelector and Parameters are mutually exclusive. 

I haven't tried generating TS from the definitions here but I've used [this library](https://github.com/bcherny/json-schema-to-typescript#readme) on other schemas before and it's worked well.

**Reply by ebisbe** [reply]
That library is the one I found. Glad it works
